vcd/resource_vcd_vapp_nat_rules_test.go

```diff
@@ -0,0 +1,369 @@
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.vm_nic_id", "0"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.protocol", "TCP_UDP"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.forward_to_port", "80"),
+
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.external_port", "11"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.vm_nic_id", "0"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.forward_to_port", "1112"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.protocol", "TCP"),
+				),
+			},
+			// we can reuse importStateVappFirewallRuleObject as import is the same
+			resource.TestStep{ // Step 1 - resource import
+				ResourceName:            "vcd_vapp_nat_rules.imported",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateIdFunc:       importStateVappFirewallRuleObject(testConfig, vappName, vappNetworkName),
+				ImportStateVerifyIgnore: []string{"enable_ip_masquerade", "network_id", "org", "vdc"},
+			},
+			resource.TestStep{ // Step 2 - resource import by ID
+				ResourceName:            "vcd_vapp_nat_rules.imported2",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateIdFunc:       importStateVappFirewallRuleById(testConfig, resourceName),
+				ImportStateVerifyIgnore: []string{"enable_ip_masquerade", "network_id", "org", "vdc"},
+			},
+			resource.TestStep{ // Step 3 - update
+				Config: configTextForUpdate,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVcdVappNatRulesExists(resourceName, 2),
+					resource.TestCheckResourceAttr(resourceName, "rule.0.mapping_mode", "manual"),
+					resource.TestCheckResourceAttr(resourceName, "rule.0.external_ip", "10.10.102.14"),
+					//resource.TestCheckResourceAttr(resourceName, "rule.0.mapping_mode", "automatic"),
+					resource.TestCheckResourceAttr(resourceName, "rule.0.vm_nic_id", "0"),
+
+					resource.TestCheckResourceAttr(resourceName, "rule.1.mapping_mode", "automatic"),
+					resource.TestCheckResourceAttr(resourceName, "rule.1.vm_nic_id", "0"),
+
+					testAccCheckVcdVappNatRulesExists(resourceName+"2", 2),
+					resource.TestCheckResourceAttr(resourceName+"2", "enable_ip_masquerade", "false"),
+
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.external_port", "11"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.vm_nic_id", "0"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.forward_to_port", "1112"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.0.protocol", "TCP"),
+
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.external_port", "222"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.vm_nic_id", "0"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.protocol", "UDP"),
+					resource.TestCheckResourceAttr(resourceName+"2", "rule.1.forward_to_port", "800"),
+				),
+			},
+		},
```

Comment:
Can we have a step which removes NAT rules, but leaves vApp/VMs in place. That way we could check that rules are really deleted because I don't see other way to check that delete works.

Reply:
Explicitly delete is verified in govcd

Reply:
I can add check though..

Reply:
Let others chime in - I vote for it. govcd test doesn't guarantee this doesn't fail.

cc @lvirbalas , @dataclouder

Reply:
added